File: cli/internal/connectapi/notifications.go
package connectapi

import (
	"context"
	"errors"
	"fmt"

	"connectrpc.com/connect"
	"hmans.de/chatto/internal/core"
	apiv1 "hmans.de/chatto/internal/pb/chatto/api/v1"
	corev1 "hmans.de/chatto/internal/pb/chatto/core/v1"
)

const (
	defaultNotificationLimit = 50
	maxNotificationLimit     = 100
)

type notificationService struct {
	api *API
}

func (s *notificationService) ListNotifications(ctx context.Context, req *connect.Request[apiv1.ListNotificationsRequest]) (*connect.Response[apiv1.ListNotificationsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	return s.notificationPage(ctx, caller.UserID, req.Msg.GetPage(), nil)
}

func (s *notificationService) ListRoomNotifications(ctx context.Context, req *connect.Request[apiv1.ListRoomNotificationsRequest]) (*connect.Response[apiv1.ListRoomNotificationsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	notifications, err := s.api.core.GetRoomNotificationsForMember(ctx, caller.UserID, req.Msg.GetRoomId())
	if err != nil {
		return nil, connectError(err)
	}
	page, err := s.notificationPageFromList(ctx, notifications, req.Msg.GetPage())
	if err != nil {
		return nil, err
	}
	return connect.NewResponse(&apiv1.ListRoomNotificationsResponse{
		Items:      page.Msg.GetItems(),
		ServerName: page.Msg.GetServerName(),
		Page:       page.Msg.GetPage(),
	}), nil
}

func (s *notificationService) HasNotifications(ctx context.Context, _ *connect.Request[apiv1.HasNotificationsRequest]) (*connect.Response[apiv1.HasNotificationsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	has, err := s.api.core.HasUnreadNotifications(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.HasNotificationsResponse{HasNotifications: has}), nil
}

func (s *notificationService) ListNotificationCounts(ctx context.Context, _ *connect.Request[apiv1.ListNotificationCountsRequest]) (*connect.Response[apiv1.ListNotificationCountsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	notifications, err := s.api.core.GetNotifications(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	countsByRoom := make(map[string]int32)
	for _, notification := range notifications {
		roomID := notificationTargetRoomID(notification)
		if roomID == "" {
			continue
		}
		countsByRoom[roomID]++
	}
	roomCounts := make([]*apiv1.RoomNotificationCount, 0, len(countsByRoom))
	for roomID, count := range countsByRoom {
		roomCounts = append(roomCounts, &apiv1.RoomNotificationCount{
			RoomId:     roomID,
			TotalCount: count,
		})
	}
	return connect.NewResponse(&apiv1.ListNotificationCountsResponse{RoomCounts: roomCounts}), nil
}

func (s *notificationService) DismissNotification(ctx context.Context, req *connect.Request[apiv1.DismissNotificationRequest]) (*connect.Response[apiv1.DismissNotificationResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	if req.Msg.NotificationId == "" {
		return nil, invalidArgument("notification_id is required")
	}
	dismissed, err := s.api.core.DismissNotification(ctx, caller.UserID, req.Msg.NotificationId)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.DismissNotificationResponse{Dismissed: dismissed}), nil
}

func (s *notificationService) DismissAllNotifications(ctx context.Context, _ *connect.Request[apiv1.DismissAllNotificationsRequest]) (*connect.Response[apiv1.DismissAllNotificationsResponse], error) {
	caller, err := requireCaller(ctx)
	if err != nil {
		return nil, err
	}
	count, err := s.api.core.DismissAllNotifications(ctx, caller.UserID)
	if err != nil {
		return nil, connectError(err)
	}
	return connect.NewResponse(&apiv1.DismissAllNotificationsResponse{DismissedCount: int32(count)}), nil
}

func (s *notificationService) notificationPage(ctx context.Context, userID string, pageRequest *apiv1.PageRequest, matches func(*corev1.Notification) bool) (*connect.Response[apiv1.ListNotificationsResponse], error) {
	notifications, err := s.api.core.GetNotifications(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}

	filtered := notifications
	if matches != nil {
		filtered = make([]*corev1.Notification, 0, len(notifications))
		for _, notification := range notifications {
			if matches(notification) {
				filtered = append(filtered, notification)
			}
		}
	}

	return s.notificationPageFromList(ctx, filtered, pageRequest)
}

func (s *notificationService) notificationPageFromList(ctx context.Context, notifications []*corev1.Notification, pageRequest *apiv1.PageRequest) (*connect.Response[apiv1.ListNotificationsResponse], error) {
	limitVal, offsetVal := apiPagination(pageRequest, defaultNotificationLimit, maxNotificationLimit)
	page, totalCount, hasMore := paginateNotifications(notifications, limitVal, offsetVal)
	items := make([]*apiv1.NotificationItem, 0, len(page))
	for _, notification := range page {
		item, err := s.apiNotificationItem(ctx, notification)
		if err != nil {
			return nil, err
		}
		if item != nil {
			items = append(items, item)
		}
	}

	response := s.emptyPage(ctx)
	response.Items = items
	response.Page = apiPageInfo(totalCount, hasMore)
	return connect.NewResponse(response), nil
}

func (s *notificationService) emptyPage(ctx context.Context) *apiv1.ListNotificationsResponse {
	name := "Chatto"
	if cm := s.api.core.ConfigManager(); cm != nil {
		if configuredName, err := cm.GetEffectiveServerName(ctx); err == nil && configuredName != "" {
			name = configuredName
		}
	}
	return &apiv1.ListNotificationsResponse{
		Items:      []*apiv1.NotificationItem{},
		ServerName: name,
	}
}

func (s *notificationService) apiNotificationItem(ctx context.Context, notification *corev1.Notification) (*apiv1.NotificationItem, error) {
	if notification == nil {
		return nil, nil
	}

	actor, err := s.notificationActor(ctx, notification.GetActorId())
	if err != nil {
		return nil, err
	}
	item := &apiv1.NotificationItem{
		Id:        notification.GetId(),
		CreatedAt: notification.GetCreatedAt(),
		Actor:     actor,
		Summary:   notificationSummary(actor, notification),
	}

	switch payload := notification.GetNotification().(type) {
	case *corev1.Notification_DmMessage:
		item.Kind = &apiv1.NotificationItem_DirectMessage{
			DirectMessage: &apiv1.DirectMessageNotification{
				RoomId:  payload.DmMessage.GetRoomId(),
				EventId: payload.DmMessage.GetEventId(),
			},
		}
	case *corev1.Notification_Mention:
		room, err := s.notificationRoom(ctx, payload.Mention.GetRoomId())
		if err != nil {
			return nil, err
		}
		mention := &apiv1.MentionNotification{
			Room:    room,
			EventId: payload.Mention.GetEventId(),
		}
		if threadID := payload.Mention.GetInThread(); threadID != "" {
			mention.ThreadRootEventId = &threadID
		}
		item.Kind = &apiv1.NotificationItem_Mention{Mention: mention}
	case *corev1.Notification_Reply:
		room, err := s.notificationRoom(ctx, payload.Reply.GetRoomId())
		if err != nil {
			return nil, err
		}
		reply := &apiv1.ReplyNotification{
			Room:        room,
			EventId:     payload.Reply.GetEventId(),
			InReplyToId: payload.Reply.GetInReplyToId(),
		}
		if threadID := payload.Reply.GetInThread(); threadID != "" {
			reply.ThreadRootEventId = &threadID
		}
		item.Kind = &apiv1.NotificationItem_Reply{Reply: reply}
	case *corev1.Notification_RoomMessage:
		room, err := s.notificationRoom(ctx, payload.RoomMessage.GetRoomId())
		if err != nil {
			return nil, err
		}
		item.Kind = &apiv1.NotificationItem_RoomMessage{
			RoomMessage: &apiv1.RoomMessageNotification{
				Room:    room,
				EventId: payload.RoomMessage.GetEventId(),
			},
		}
	default:
		return nil, connectError(fmt.Errorf("unknown notification type %T", notification.GetNotification()))
	}

	return item, nil
}

func (s *notificationService) notificationActor(ctx context.Context, userID string) (*apiv1.UserProfile, error) {
	if userID == "" {
		return nil, nil
	}
	user, err := s.api.core.GetUser(ctx, userID)
	if err != nil {
		if errors.Is(err, core.ErrNotFound) {
			return nil, nil
		}
		return nil, connectError(err)
	}
	presence, err := s.api.core.GetUserPresence(ctx, userID)
	if err != nil {
		return nil, connectError(err)
	}
	actor := &apiv1.UserProfile{
		User: &apiv1.User{
			Id:          user.GetId(),
			Login:       user.GetLogin(),
			DisplayName: user.GetDisplayName(),
			Deleted:     user.GetDeleted(),
		},
		PresenceStatus: corePresenceStatusToAPI(presence),
		CustomStatus:   coreCustomStatusToAPI(user.GetCustomStatus()),
	}
	if avatarURL, err := s.api.core.GetUserAvatarURL(ctx, userID, nil, nil, ""); err != nil {
		return nil, connectError(err)
	} else if avatarURL != "" {
		actor.User.AvatarUrl = stringPtr(s.api.absolutizeAssetURL(ctx, avatarURL))
	}
	return actor, nil
}

func (s *notificationService) notificationRoom(ctx context.Context, roomID string) (*apiv1.NotificationRoom, error) {
	if roomID == "" {
		return nil, nil
	}
	room, err := s.api.core.FindRoomByID(ctx, roomID)
	if err != nil {
		return nil, connectError(err)
	}
	return &apiv1.NotificationRoom{
		Id:   room.GetId(),
		Name: room.GetName(),
	}, nil
}

func notificationSummary(actor *apiv1.UserProfile, notification *corev1.Notification) string {
	actorName := ""
	if actor != nil && actor.GetUser() != nil {
		actorName = actor.GetUser().GetDisplayName()
	}
	switch notification.GetNotification().(type) {
	case *corev1.Notification_DmMessage:
		if actorName == "" {
			return "New message"
		}
		return fmt.Sprintf("%s sent you a message", actorName)
	case *corev1.Notification_Mention:
		if actorName == "" {
			return "You were mentioned"
		}
		return fmt.Sprintf("%s mentioned you", actorName)
	case *corev1.Notification_Reply:
		if actorName == "" {
			return "New reply to your message"
		}
		return fmt.Sprintf("%s replied to your message", actorName)
	case *corev1.Notification_RoomMessage:
		if actorName == "" {
			return "New message"
		}
		return fmt.Sprintf("%s posted a message", actorName)
	default:
		return "New notification"
	}
}

func notificationTargetRoomID(notification *corev1.Notification) string {
	if notification == nil {
		return ""
	}
	switch payload := notification.GetNotification().(type) {
	case *corev1.Notification_DmMessage:
		return payload.DmMessage.GetRoomId()
	case *corev1.Notification_Mention:
		return payload.Mention.GetRoomId()
	case *corev1.Notification_Reply:
		return payload.Reply.GetRoomId()
	case *corev1.Notification_RoomMessage:
		return payload.RoomMessage.GetRoomId()
	default:
		return ""
	}
}

func paginateNotifications(notifications []*corev1.Notification, limit, offset int) ([]*corev1.Notification, int, bool) {
	total := len(notifications)
	if offset >= total {
		return []*corev1.Notification{}, total, false
	}
	end := offset + limit
	if end > total {
		end = total
	}
	return notifications[offset:end], total, end < total
}

File: apps/frontend/src/lib/api/notifications.ts
import { createClient } from '@connectrpc/connect';
import { createConnectTransport } from '@connectrpc/connect-web';
import { NotificationService } from '$lib/pb/chatto/api/v1/notifications_connect';
import type {
  ListRoomNotificationsResponse,
  ListNotificationsResponse,
  NotificationItem as APINotificationItem
} from '$lib/pb/chatto/api/v1/notifications_pb';
import type { UserPresenceSummary as APIUserPresenceSummary } from '$lib/pb/chatto/api/v1/users_pb';
import { PresenceStatus as APIPresenceStatus } from '$lib/pb/chatto/api/v1/presence_pb';
import { PresenceStatus } from '$lib/render/types';

export type NotificationAPIConfig = {
  baseUrl: string;
  bearerToken: string | null;
};

export type NotificationActor = {
  id: string;
  login: string;
  displayName: string;
  deleted: boolean;
  avatarUrl?: string | null;
  presenceStatus: PresenceStatus;
  customStatus?: {
    emoji: string;
    text: string;
    expiresAt?: string | null;
  } | null;
};

export const NotificationItemKind = {
  DirectMessage: 'directMessage',
  Mention: 'mention',
  Reply: 'reply',
  RoomMessage: 'roomMessage'
} as const;

export type NotificationItemKind = (typeof NotificationItemKind)[keyof typeof NotificationItemKind];

export type DirectMessageNotificationItem = {
  kind: typeof NotificationItemKind.DirectMessage;
  id: string;
  createdAt: string;
  actor?: NotificationActor | null;
  summary: string;
  room: { id: string };
};

export type MentionNotificationItem = {
  kind: typeof NotificationItemKind.Mention;
  id: string;
  createdAt: string;
  actor?: NotificationActor | null;
  summary: string;
  mentionRoom: { id: string; name: string } | null;
  mentionEventId: string;
  mentionInThread?: string | null;
};

export type ReplyNotificationItem = {
  kind: typeof NotificationItemKind.Reply;
  id: string;
  createdAt: string;
  actor?: NotificationActor | null;
  summary: string;
  replyRoom: { id: string; name: string } | null;
  replyEventId: string;
  inReplyToId: string;
  replyInThread?: string | null;
};

export type RoomMessageNotificationItem = {
  kind: typeof NotificationItemKind.RoomMessage;
  id: string;
  createdAt: string;
  actor?: NotificationActor | null;
  summary: string;
  roomMsgRoom: { id: string; name: string } | null;
  roomMsgEventId: string;
};

export type NotificationItem =
  | DirectMessageNotificationItem
  | MentionNotificationItem
  | ReplyNotificationItem
  | RoomMessageNotificationItem;

export type NotificationPage = {
  items: NotificationItem[];
  totalCount: number;
  hasMore: boolean;
  serverName: string | null;
};

export function createNotificationAPI(config: NotificationAPIConfig) {
  const transport = createConnectTransport({
    baseUrl: config.baseUrl,
    useBinaryFormat: true
  });
  const client = createClient(NotificationService, transport);
  const headers = () =>
    config.bearerToken ? { Authorization: `Bearer ${config.bearerToken}` } : undefined;

  return {
    async listNotifications(limit = 50, offset = 0): Promise<NotificationPage> {
      return notificationPage(
        await client.listNotifications({ page: { limit, offset } }, { headers: headers() })
      );
    },

    async listRoomNotifications(roomId: string, limit = 1, offset = 0): Promise<NotificationPage> {
      return notificationPage(
        await client.listRoomNotifications({ roomId, page: { limit, offset } }, { headers: headers() })
      );
    },

    async hasNotifications(): Promise<boolean> {
      return (await client.hasNotifications({}, { headers: headers() })).hasNotifications;
    },

    async listNotificationCounts(): Promise<Record<string, number>> {
      const response = await client.listNotificationCounts({}, { headers: headers() });
      return Object.fromEntries(
        response.roomCounts.map((count) => [count.roomId, count.totalCount] as const)
      );
    },

    async dismissNotification(notificationId: string): Promise<boolean> {
      return (await client.dismissNotification({ notificationId }, { headers: headers() }))
        .dismissed;
    },

    async dismissAllNotifications(): Promise<number> {
      return (await client.dismissAllNotifications({}, { headers: headers() })).dismissedCount;
    }
  };
}

export type NotificationAPI = ReturnType<typeof createNotificationAPI>;

function notificationPage(response: ListNotificationsResponse | ListRoomNotificationsResponse): NotificationPage {
  return {
    items: response.items.flatMap((item) => {
      const mapped = notificationItem(item);
      return mapped ? [mapped] : [];
    }),
    totalCount: Number(response.page?.totalCount ?? 0),
    hasMore: response.page?.hasMore ?? false,
    serverName: response.serverName || null
  };
}

function notificationItem(item: APINotificationItem): NotificationItem | null {
  const base = {
    id: item.id,
    createdAt: item.createdAt?.toDate().toISOString() ?? new Date(0).toISOString(),
    actor: notificationActor(item.actor),
    summary: item.summary
  };

  switch (item.kind.case) {
    case 'directMessage':
      return {
        kind: NotificationItemKind.DirectMessage,
        ...base,
        room: { id: item.kind.value.roomId }
      };
    case 'mention':
      return {
        kind: NotificationItemKind.Mention,
        ...base,
        mentionRoom: item.kind.value.room
          ? { id: item.kind.value.room.id, name: item.kind.value.room.name }
          : null,
        mentionEventId: item.kind.value.eventId,
        mentionInThread: item.kind.value.threadRootEventId ?? null
      };
    case 'reply':
      return {
        kind: NotificationItemKind.Reply,
        ...base,
        replyRoom: item.kind.value.room
          ? { id: item.kind.value.room.id, name: item.kind.value.room.name }
          : null,
        replyEventId: item.kind.value.eventId,
        inReplyToId: item.kind.value.inReplyToId,
        replyInThread: item.kind.value.threadRootEventId ?? null
      };
    case 'roomMessage':
      return {
        kind: NotificationItemKind.RoomMessage,
        ...base,
        roomMsgRoom: item.kind.value.room
          ? { id: item.kind.value.room.id, name: item.kind.value.room.name }
          : null,
        roomMsgEventId: item.kind.value.eventId
      };
    default:
      return null;
  }
}

function notificationActor(actor: APIUserPresenceSummary | undefined): NotificationActor | null {
  const summary = actor?.user;
  if (!actor || !summary) return null;
  return {
    id: summary.id,
    login: summary.login,
    displayName: summary.displayName,
    deleted: summary.deleted,
    avatarUrl: summary.avatarUrl ?? null,
    presenceStatus: apiPresenceStatus(actor.presenceStatus),
    customStatus: actor.customStatus
      ? {
          emoji: actor.customStatus.emoji,
          text: actor.customStatus.text,
          expiresAt: actor.customStatus.expiresAt?.toDate().toISOString() ?? null
        }
      : null
  };
}

function apiPresenceStatus(status: APIPresenceStatus): PresenceStatus {
  switch (status) {
    case APIPresenceStatus.AWAY:
      return PresenceStatus.Away;
    case APIPresenceStatus.DO_NOT_DISTURB:
      return PresenceStatus.DoNotDisturb;
    case APIPresenceStatus.ONLINE:
      return PresenceStatus.Online;
    case APIPresenceStatus.OFFLINE:
    case APIPresenceStatus.UNSPECIFIED:
    default:
      return PresenceStatus.Offline;
  }
}
